article.md

```diff
@@ -0,0 +1,265 @@
+        hashed_msg : Uint256){
+
+    let (starknet_address_h, starknet_address_l) = split_felt(starknet_address);
+    
+    let (encoded_data : Uint256*) = alloc();
+    assert encoded_data[0] = Uint256(TYPE_HASH_LOW, TYPE_HASH_HIGH);
+    assert encoded_data[1] = Uint256(starknet_address_l, starknet_address_h);
+    let (data_hash) = keccak_uint256s_bigend{keccak_ptr=keccak_ptr}(2, encoded_data);
+
+    let prefix = PREFIX;
+    let (w0, prefix) = add_prefix(domain_hash.high, prefix);
+    let (w1, prefix) = add_prefix(domain_hash.low, prefix);
+    let (w2, prefix) = add_prefix(data_hash.high, prefix);
+    let (w3, overflow) = add_prefix(data_hash.low, prefix);
+    
+    let (signable_bytes : felt*) = alloc();
+    
+    // w0, w1, w2, w3, and overflow need to be split into 64 bit chunks in little endian format
+    populate_array_little_end(w0, w1, w2, w3, overflow, signable_bytes);
+
+    // final value is 66 bytes long 
+    let (res) = keccak_bigend{keccak_ptr=keccak_ptr}(signable_bytes, 66);
+    
+    return (hashed_msg = res);  
+}
+```
+All that's left to do is call `verify_eth_signature_uint256` in our `assert_valid_eth_signature` function.
+```cairo
+    verify_eth_signature_uint256{keccak_ptr=keccak_ptr}(hash_uint, r, s, v, eth_address);
+
+    finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+
+    return ();
+}
+```
+We need to call `finaliza_keccak` because of the internal batching done.
+inside it. Failing to call `finalize_keccak()` will make the keccak function unsound - the prover will be able to choose any value as the keccak's result.
+## Finalizing
+If everything matches, no exception will be thrown and `save_connected_addresses` function will be called, which will bind our two addresses together. <br/>
+
+```cairo
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.math import assert_nn
+
+
+@storage_var
+func storage(eth_address : felt, starknet_address : felt) -> (exists: felt){
+}
+
+func save_connected_addresses{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+}(eth_address: felt, starknet_address: felt){
+
+    storage.write(eth_address, starknet_address, 1);
+    return ();
+}
+
+func are_addresses_connected{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+}(eth_address: felt, starknet_address: felt) -> (res : felt){
+
+    let (res) = storage.read(eth_address, starknet_address);
+    return (res = res);
+}
+```
+Bare in mind that because of our implementation of `map.cairo` we won't be able to, for example, list all the Starknet addresses connected to the Ethereum address or vice versa. All we can do is check whether some two addresses are connected or not. If you want some more advanced functions, you'd need to change how the addresses are bound - for example, instead of <br/>`func storage(eth_address : felt, starknet_address : felt)`
```

Comment:
**Bear** in mind that we won't be able to list all the Starknet addresses connected to the Ethereum address or vice versa.